query/workmanager.go

```diff
@@ -0,0 +1,455 @@
+// among available workers, orchestrating where to send them.
+type WorkManager struct {
+	cfg *Config
+
+	// newBatches is a channel where new batches of queries will be sent to
+	// the workDispatcher.
+	newBatches chan *batch
+
+	// jobResults is the common channel where results from queries from all
+	// workers will be sent.
+	jobResults chan *jobResult
+
+	quit chan struct{}
+	wg   sync.WaitGroup
+}
+
+// Compile time check to ensure WorkManager satisfies the Dispatcher interface.
+var _ Dispatcher = (*WorkManager)(nil)
+
+// New returns a new WorkManager with the regular worker implementation.
+func New(cfg *Config) *WorkManager {
+	return &WorkManager{
+		cfg:        cfg,
+		newBatches: make(chan *batch),
+		jobResults: make(chan *jobResult),
+		quit:       make(chan struct{}),
+	}
+}
+
+// Start starts the WorkManager.
+func (w *WorkManager) Start() error {
+
+	// Get a peer subscription.
+	peersConnected, cancel, err := w.cfg.ConnectedPeers()
+	if err != nil {
+		return fmt.Errorf("Unable to get connected peers: %v", err)
+	}
+
+	w.wg.Add(1)
+	go func() {
+		defer w.wg.Done()
+		defer cancel()
+
+		w.workDispatcher(peersConnected)
+	}()
+
+	return nil
+}
+
+// Stop stops the WorkManager and all underlying goroutines.
+func (w *WorkManager) Stop() error {
+	close(w.quit)
+	w.wg.Wait()
+
+	return nil
+}
+
+// workDispatcher receives batches of queries to be performed from external
+// callers, and dispatches these to active workers.  It makes sure to
+// prioritize the queries in the order they come in, such that early queries
+// will be attemped completed first.
+func (w *WorkManager) workDispatcher(peersConnected <-chan Peer) {
+	// Init a work queue which will be used to sort the incoming queries in
+	// a first come first served fashion. We use a heap structure such
+	// that we can efficiently put failed queries back in the queue.
+	work := &workQueue{}
+	heap.Init(work)
+
+	type batchProgress struct {
+		timeout <-chan time.Time
+		rem     int
+		errChan chan error
+	}
+
+	// We set up a batch index counter to keep track of batches that still
+	// have queries in flight. This lets us track when all queries for a
+	// batch has been finished, and return an (non-)error to the caller.
+	batchIndex := uint64(0)
+	currentBatches := make(map[uint64]*batchProgress)
+
+	// When the work dispatcher exits, we'll loop through the remaining
+	// batches and send on their error channel.
+	defer func() {
+		for _, b := range currentBatches {
+			b.errChan <- ErrWorkManagerShuttingDown
+		}
+	}()
+
+	// We set up a counter that we'll increase with each incoming query,
+	// and will serve as the priority of each. In addition we map each
+	// query to the batch they are part of.
+	queryIndex := uint64(0)
+	currentQueries := make(map[uint64]uint64)
+
+	workers := make(map[string]*activeWorker)
+
+Loop:
+	for {
+		// If the work queue is non-empty, we'll take out the first
+		// element in order for distributing it to a worker.
```

Comment:
to distribute it to a worker

Reply:
`ErrChans` are buffered with 1, and always safe to send on once.